cmd/activator/main.go

```diff
@@ -145,8 +146,15 @@ func main() {
 		logger.Fatalw("Error building serving clientset", zap.Error(err))
 	}
 
-	if err := version.CheckMinimumVersion(kubeClient.Discovery()); err != nil {
-		logger.Fatalf("Version check failed: %v", err)
+	// We sometimes startup faster than we can reach kube-api. Poll on failure to prevent us terminating
+	if err := wait.PollImmediate(time.Second, 60*time.Second, func() (bool, error) {
+		if err = version.CheckMinimumVersion(kubeClient.Discovery()); err != nil {
+			logger.Errorw("Failed to get k8s version", zap.Error(err))
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		logger.Fatalw("Timed out attempting to get k8s version", zap.Error(err))
```

Comment:
I think you want err on line 151 be `err :=`, otherwise they'll conmingle.

Reply:
Thinking about it, you probably want to create an err variable outside of the poller and report the last error reported by the retrying function, like:

```
var err error
wait.PollImmediate(..., func() {
  err = ...
})
```

That's usually more helpful in the end than just reporting the timeout error.

Reply:
Heh :) here's what I had: https://github.com/knative/serving/compare/master...vagababov:wait-for-start?expand=1

Reply:
Happy to rework it, but practically I don't think it should be much of a difference - the error var masking shouldnt be an issue since the poll method is blocking. I agree thats not super pretty though.

Re: logging - this logs the error on every failure so a user should see both that and then eventually a timeout. I figured this is nicer than silence for a while which id expect a user to wonder why the service is hanging.

Reply:
Ehr, you're right. The error logging would've printed the inner error all along. The old version would've been fine as well.